File: src/1512_number_of_good_pairs/number_of_good_pairs.py
from typing import List
from collections import defaultdict


class Solution:
    def numIdenticalPairs(self, nums: List[int]) -> int:
        count = defaultdict(int)
        res = 0
        for num in nums:
            res += count[num]
            count[num] += 1
        return res


def test():
    sol = Solution()
    print(sol.numIdenticalPairs([1, 2, 3, 1, 1, 3]), "| 4")
    print(sol.numIdenticalPairs([1, 1, 1, 1]), "| 6")
    print(sol.numIdenticalPairs([1, 2, 3]), "| 0")
    print(sol.numIdenticalPairs([]), "| 0")
    print(sol.numIdenticalPairs([2, 2, 2, 2, 2]), "| 10")


if __name__ == "__main__":
    test()

File: src/1512_number_of_good_pairs/number_of_good_pairs.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
   public:
    int numIdenticalPairs(vector<int>& nums) {
        unordered_map<int, int> count;
        int res = 0;
        for (int num : nums) {
            res += count[num];
            count[num]++;
        }
        return res;
    }
};

int main() {
    Solution sol;
    vector<int> t1 = {1, 2, 3, 1, 1, 3};
    cout << sol.numIdenticalPairs(t1) << " | 4" << endl;
    vector<int> t2 = {1, 1, 1, 1};
    cout << sol.numIdenticalPairs(t2) << " | 6" << endl;
    vector<int> t3 = {1, 2, 3};
    cout << sol.numIdenticalPairs(t3) << " | 0" << endl;
    vector<int> t4 = {};
    cout << sol.numIdenticalPairs(t4) << " | 0" << endl;
    vector<int> t5 = {2, 2, 2, 2, 2};
    cout << sol.numIdenticalPairs(t5) << " | 10" << endl;
    return 0;
}
